File: object builder/object_builder.py
# -*- coding=Utf-8 -*-
from datetime import date
from nota_fiscal import NotaFiscal

class ObjectBuilder(object):

	def __init__(self):
		self.__titulo = None
		self.__itens = None
		self.__descricao = None
		self.__data_vencimento = None

	def com_titulo(self, titulo):
		self.__titulo = titulo
		return self

	def com_itens(self, itens):
		self.__itens = itens
		return self

	def com_descricao(self, descricao):
		self.__descricao = descricao
		return self

	def com_data_vencimento(self, data):
		self.__data_vencimento = data
		return self

	def constroi(self):
		if self.__titulo is None:
			raise Exception("Titulo não pode estar vazio")

		if self.__itens is None:
			raise Exception("Itens não pode sem vazio")

		if self.__descricao is None:
			self.__descricao = ''

		if self.__data_vencimento is None:
			self.__data_vencimento = date.today()

		return NotaFiscal(titulo=self.__titulo, itens=self.__itens, descricao=self.__descricao, data_vencimento=self.__data_vencimento)

File: java/workspace/design-pattern/src/object_builder/ObjectBuilder.java
package object_builder;

import java.util.Date;

public class ObjectBuilder {

	private String titulo;
	private String codigo;
	private Date date;
	private String descricao;
	
	public ObjectBuilder comTitulo(String titulo){
		this.titulo = titulo;
		return this;
	}
	
	public ObjectBuilder comCodigo(String codigo){
		this.codigo = codigo;
		return this;
	}
	
	public ObjectBuilder comDate(Date date){
		this.date = date;
		return this;
	}
	
	public ObjectBuilder comDescricao(String descricao){
		this.descricao = descricao;
		return this;
	}
	
	public NotaFiscal construir() throws Exception{
		if(titulo == null || titulo.isEmpty()){
			throw new Exception("Titulo não pode ser em branco");
		}
		if(codigo == null || codigo.isEmpty()){
			throw new Exception("codigo não pode ser em branco");
		}
		if(date == null){
			this.date = new Date();
		}
		if(descricao == null ){
			this.descricao = "";
		}
		return new NotaFiscal(titulo, codigo, date, descricao);
	}
}
